### PR TITLE
Add block-nav package

### PR DESCRIPTION
**Pull request by nixin72**
### Brief summary of what the package does
This package allows the user to quickly navigate around code blocks by:
- jumping between lines of code with the same level of indentation
- jumping forwards/backwards to lines with a deeper/shallower level of indentation.

This gif demonstrates use of the package:
ezgif com-optimize

### Direct link to the package repository
https://github.com/nixin72/block-nav.el

### Your association with the package
Author

### Relevant communications with the upstream package maintainer

**None needed**

### Checklist

Please confirm with `x`:

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses).
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [x] My elisp byte-compiles cleanly
- [x] `M-x checkdoc` is happy with my docstrings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [ ] I have confirmed some of these without doing them

### CODE REVIEW

**Comment by alphapapa** [comment]
Are you sure that this functionality is not already covered by the built-in commands, like `forward-sexp`, `forward-list`, `down-list`, `backward-up-list`, etc?

**Comment by nixin72** [comment]
Not as far as I can tell. For example, `backward-sexpr` works very similarly to `block-nav-previous-block` if using it to move backwards across top-level definitions, but if I were to use it here:
```
(define function-1
  (lambda (x) x))

;; comment

(define function-2
  (lambda (x) x)) 
```
And my point is at the opening parentheses to the lambda for `function-2`, `backward-sexpr` would move to the symbol `function-2`, then `define`, but it won't break out of the s-expr. Using `block-nav-previous-block` would jump from the beginning of the lambda for `function-2` to the beginning of the lambda for `function-1`. This is not something that I could find a built-in command that can do. 
(optionally, if you have the `block-nav-skip-shallower` variable set to nil, you'll be moved to the opening parentheses of the second `define`).

I also don't see a way to change whether these functions will skip comments. If I use `backward-sexpr` with point at the beginning of the second `define`, it'll skip over that comment. `block-nav-previous-block` gives the option to stop at that comment. 

I think that this set of functions is quite different from what the builtins provide. I provided examples here for `backward-sexpr`, but in my testing the functionality appears to be quite different for others as well.

**Comment by riscy** [comment]
Just noting this seems like an inverted (?) version of https://github.com/melpa/melpa/pull/7159

**Comment by nixin72** [comment]
Yea, they were inspired by the same reddit post @riscy: https://www.reddit.com/r/emacs/comments/izwrlf/anyone_using_spatial_navigation_in_emacs/

**Comment by nixin72** [comment]
I made a post about this on r/emacs the following day showing the first POC: https://www.reddit.com/r/emacs/comments/j0vtqc/i_really_liked_that_spacial_navigation_shared_the/?utm_source=share&utm_medium=web2x&context=3

**Comment by alphapapa** [comment]
> > Are you sure that this functionality is not already covered by the built-in commands, like forward-sexp, forward-list, down-list, backward-up-list, etc?

> And my point is at the opening parentheses to the lambda for function-2, backward-sexpr would move to the symbol function-2, then define, but it won't break out of the s-expr.

The last command I mentioned, `backward-up-list`, does that.  i.e. (with point represented by `|`):

```el
(define function-1
  (lambda (x) x))

;; comment

(define function-2
  |(lambda (x) x)) 
```
`C-M-u`
```el
(define function-1
  (lambda (x) x))

;; comment

|(define function-2
  (lambda (x) x)) 
```

**Comment by riscy** [comment]
Just my two cents -- I think one place where this works and backward-up-list doesn't is in Python code which (for better or worse) has no structural delimiters that Emacs can recognize as lists.

Quick first pass in the meantime (I think you need to add a `defgroup` -- I left a similar comment in the related PR):

### block-nav.el ###

byte-compile (using Emacs 27.1):
```
block-nav.el:46:1:Warning: defcustom for `block-nav-center-after-scroll' fails
    to specify type
block-nav.el:46:1:Warning: defcustom for `block-nav-center-after-scroll' fails
    to specify containing group
block-nav.el:46:1:Warning: defcustom for `block-nav-center-after-scroll' fails
    to specify type
block-nav.el:46:1:Warning: defcustom for `block-nav-center-after-scroll' fails
    to specify containing group
block-nav.el:48:1:Warning: defcustom for `block-nav-move-skip-shallower' fails
    to specify type
block-nav.el:48:1:Warning: defcustom for `block-nav-move-skip-shallower' fails
    to specify containing group
block-nav.el:48:1:Warning: defcustom for `block-nav-move-skip-shallower' fails
    to specify type
block-nav.el:48:1:Warning: defcustom for `block-nav-move-skip-shallower' fails
    to specify containing group
block-nav.el:50:1:Warning: defcustom for `block-nav-skip-comment' fails to
    specify type
block-nav.el:50:1:Warning: defcustom for `block-nav-skip-comment' fails to
    specify containing group
block-nav.el:50:1:Warning: defcustom for `block-nav-skip-comment' fails to
    specify type
block-nav.el:50:1:Warning: defcustom for `block-nav-skip-comment' fails to
    specify containing group
```

checkdoc (using version 0.6.2):
```
Warning (emacs): 
block-nav.el:63: First sentence should end with punctuation
```

package-lint (using version 20200906.512):
- No issues!

Other possible lints:
- block-nav.el#L66: Could this point/point-at-eol comparison use `eolp`?
- block-nav.el#L7: Prefer `https` over `http` if possible ([why?](https://news.ycombinator.com/item?id=22933774))
- block-nav.el#L26: Prefer `https` over `http` if possible ([why?](https://news.ycombinator.com/item?id=22933774))

**Comment by nixin72** [comment]
Okay, I've fixed all the errors @riscy. 

- Added a `defgroup` for the file
- Added `:type` and `:group` for each of the `defcustom`s
- Replaced my end of line check with `eolp`
- Replaced instances of `http` with `https`

And now 
- `byte-compile-file` gives no warnings 
- `checkdoc` gives no warnings 
- `package-lint-current-buffer` gives no warnings

**Comment by alphapapa** [comment]
> Just my two cents -- I think one place where this works and backward-up-list doesn't is in Python code which (for better or worse) has no structural delimiters that Emacs can recognize as lists.

Strictly, that's true, however Emacs tends to have this covered already.  e.g. in `python-mode` buffers, `C-M-u` is bound to:
```
C-M-u runs the command python-nav-backward-up-list (found in
python-mode-map), which is an interactive compiled Lisp function in
‘python.el’.

It is bound to C-M-u, ESC <C-up>.

(python-nav-backward-up-list &optional ARG)

Move backward out of one level of parentheses (or blocks).
With ARG, do this that many times.
A negative argument means move forward but still to a less deep spot.
This command assumes point is not in a string or comment.
```
And it does work as expected, e.g.

```python
def trash(paths, bin=None):
    """Move paths to trash bin."""

    # Instantiate bin once
    bin = TrashBin()

    for path in paths:
        |TrashedPath(path, bin=bin).trash()
```

`C-M-u`

```python
def trash(paths, bin=None):
    """Move paths to trash bin."""

    # Instantiate bin once
    bin = TrashBin()

    |for path in paths:
        TrashedPath(path, bin=bin).trash()
```

**Comment by nixin72** [comment]
How about for `block-nav-next/previous-block` though? Is there a way that you can do this:
```
def trash(paths, bin=None):
    """Move paths to trash bin."""

    # Instantiate bin once
    bin = TrashBin()

    for path in paths:
        |TrashedPath(path, bin=bin).trash()
       #^ Move from here

def fun2(something):
    if something:
        |print("Here")
       #^ to here?
```
because `block-nav-next-block` gives you the ability to skip over any blocks shallower in indentation than the current one. 

`block-nav-next-indentation-level` will also skip anything else at the current level of indentation, so 
```
def trash(paths, bin=None):
    |"""Move paths to trash bin."""
   #^ this point 
    # Instantiate bin once
    bin = TrashBin()

    for path in paths:
        |TrashedPath(path, bin=bin).trash()
       #^ would be moved to here calling `block-nav-next-indentation-level`
```

And even if combining built-in functions achieves most of this functionality, these 4 functions let you really easily bind them sensibly, so for example I use `evil-mode` and have them bound like this:
```
(define-key evil-motion-state-map "H" 'block-nav-previous-indentation-level)
(define-key evil-motion-state-map "J" 'block-nav-next-block)
(define-key evil-motion-state-map "K" 'block-nav-previous-block)
(define-key evil-motion-state-map "L" 'block-nav-next-indentation-level)
```
And I have all 4 bound to my normal directional keys. Having more than 4 functions makes this much more awkward I think. 

Plus, there's less ability to customize when dealing with the built-in functions I find. `block-nav-next-indentation-level` for example gives you the option to stop at comments or not, or `block-nav-next-block` gives you the option to stop at lines with a shallower indentation level for example. Unless I'm missing something when looking through `M-x customize`, that functionality is pretty hard to modify.

**Comment by alphapapa** [comment]
I'm not aware of any built-in commands that do those operations in a single command.

IMHO, those specific commands aren't very useful, because they require the user to 1) visually parse the code between point and the desired location to determine whether the indentation of point is the same as that at the desired location, so he can 2) select the appropriate command to move point there.  

(In the same way, I don't find repeating or numerically prefixed editing/movement commands generally useful in Vim or Emacs, because they require the user to visually count ahead rather than letting the editor handle it or doing it visually.  e.g. the difference between `5dw`/`C-u 5 M-d` and either `M-d M-d M-d M-d M-d` or `C-SPC M-f M-f M-f M-f C-w`: the prefixed version may seem simpler, but having to count can take longer and is usually more error-prone than deleting until done or marking the region to be deleted.)

So I would probably never use such a command, because there are numerous faster, more direct, and easier ways to move point, like `C-s`, or other packages like `ace-jump`.

In other words, these commands are like a new vocabulary of compound words to describe specific actions in specific circumstances, while an Emacs user already has other, simpler terms that can be combined to describe the desired results in any circumstances.  In terms of the earlier examples, it would be like adding a new `delete-the-next-five-words` command instead of doing `5dw`.

YMMV.  :)

**Comment by nixin72** [comment]
And that's fair if you don't particularly see it as being something useful. Personally, I've been using this the past week since I made it and I've been finding it very useful for quickly navigating through chunks of code, and I think it'll be even more useful when using something that has a more consistent indentation like JavaScript. 

`ace-jump` also looks pretty nice too though, I'll check that out to use in conjunction with this for jumping to things that aren't at the beginning of lines. It reminds me of the hyperlink selection in the Firefox Vimium plugin that I use all the time.

**Comment by riscy** [comment]
Thanks very much both, I think this was a valuable discussion, and probably the best thing possible to happen to a MELPA PR.  Sometimes Emacs or popular existing packages already support something we have in mind, but I think that wasn't exactly the case here.  Still, I think it took a bit of digging to extract the novelty -- even if that amounts to "creates a more consistent experience" -- and I hope that can be used to improve the Commentary and/or upstream README.
